File: agents/reviewer.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any

from core.openrouter import chat
from core.similarity import compute_similarity
from config import REVIEWER_MODEL

# Reuse PII patterns from context_simplifier
from agents.context_simplifier import (
    EMAIL_RE, PHONE_RE, AGE_RE, NAME_RE, LOCATION_RE, FILENAME_PATTERN
)

@dataclass
class ReviewResult:
    approved: bool
    confidence: float  # 0.0-1.0
    similarity_score: float
    checks: dict[str, bool]
    missing_items: list[str]
    reason: str
    suggestions: list[str]  # Cho retry hint


def _quick_checks(original: str, sanitized: str) -> tuple[bool, ReviewResult | None]:
    """Phase 1: Hard deterministic checks."""
    
    # Empty output
    if not sanitized or len(sanitized.strip()) < 5:
        return False, ReviewResult(
            approved=False, confidence=1.0, similarity_score=0.0,
            checks={"empty": False}, missing_items=["sanitized_prompt"],
            reason="Output is empty or too short", suggestions=["Check Agent 2 pipeline"]
        )
    
    # PII remaining
    pii_patterns = [EMAIL_RE, PHONE_RE, AGE_RE, NAME_RE, LOCATION_RE]
    found_pii = []
    for pattern in pii_patterns:
        if pattern.search(sanitized):
            found_pii.append(pattern.pattern[:20] + "...")
    
    if found_pii:
        return False, ReviewResult(
            approved=False, confidence=1.0, similarity_score=0.0,
            checks={"pii_clean": False}, missing_items=found_pii,
            reason="PII detected in output", suggestions=["Increase PII masking in Agent 2"]
        )
    
    # Severe semantic drift (similarity < 0.3)
    sim = compute_similarity(original, sanitized)
    if sim < 0.3:
        return False, ReviewResult(
            approved=False, confidence=0.9, similarity_score=sim,
            checks={"semantic_similarity": False}, missing_items=["core_intent"],
            reason=f"Semantic drift too severe ({sim:.2f} < 0.3)",
            suggestions=["Agent 2 over-compressed; retry with lower threshold"]
        )
    
    return True, None  # Pass quick checks


def _extract_key_elements(text: str) -> dict[str, list[str]]:
    """Extract structured elements from text for comparison."""
    
    elements = {
        "filenames": FILENAME_PATTERN.findall(text),
        "errors": re.findall(r"\b\w*Error:?\s*[^.\n]*", text, re.IGNORECASE),
        "questions": re.findall(r"[^.?!]*\?[^.?!]*", text),
        "constraints": re.findall(
            r"\b(do not|don't|must|should|never|always|keep|preserve|maintain)\b[^.?!]*[.?!]",
            text, re.IGNORECASE
        ),
        "code_symbols": re.findall(r"\b\w+\(\s*\)", text),  # function calls
        "urls_apis": re.findall(r"\b(api|endpoint|url|http)[^\s]*", text, re.IGNORECASE),
    }
    return elements


def _build_review_prompt(original: str, sanitized: str, sim_score: float) -> str:
    """Build LLM prompt for nuanced review."""
    
    orig_elems = _extract_key_elements(original)
    
    return f"""You are a strict context validation agent. Compare ORIGINAL vs SANITIZED.

ORIGINAL:
\"\"\"{original[:2000]}\"\"\"

SANITIZED (Agent 2 output):
\"\"\"{sanitized[:1500]}\"\"\"

Pre-computed similarity: {sim_score:.3f}

Extracted from original (for reference):
- Filenames: {orig_elems['filenames'][:10]}
- Errors: {[e[:50] for e in orig_elems['errors'][:5]]}
- Constraints: {orig_elems['constraints'][:5]}
- Questions: {orig_elems['questions'][:3]}

VALIDATION RULES:
1. INTENT: Does sanitized capture the core request/question? (If original asks about weather, sanitized must mention weather)
2. FILES: Are all actively referenced filenames preserved?
3. ERRORS: Are error messages/stack traces kept?
4. CONSTRAINTS: Are "do not/must/keep" requirements preserved?
5. TECHNICAL: Are API names, function calls, important symbols intact?

DECISION:
- APPROVE only if ALL critical elements preserved
- FAIL if any constraint dropped or core intent changed

Output JSON:
{{
  "approved": boolean,
  "confidence": 0.0-1.0,
  "checks": {{
    "intent_preserved": boolean,
    "files_preserved": boolean,
    "errors_preserved": boolean,
    "constraints_preserved": boolean,
    "technical_symbols_intact": boolean
  }},
  "missing_items": ["specific items Agent 2 dropped"],
  "false_positives": ["items Agent 2 kept that should be dropped"],
  "reason": "detailed explanation if FAIL",
  "retry_suggestion": "specific hint for Agent 2: raise SVT threshold / add constraint keyword check / etc."
}}"""


def _extract_json_from_response(text: str) -> dict:
    """Extract JSON from LLM response, stripping markdown fences and thinking tags."""
    body = text.strip()
    # Strip <think>...</think> blocks (qwen3)
    body = re.sub(r"<think>.*?</think>", "", body, flags=re.DOTALL).strip()
    # Strip markdown code fences
    if "```" in body:
        body = re.sub(r"^```(?:json)?\s*", "", body)
        body = re.sub(r"\s*```$", "", body)
        body = body.strip()
    # Find the JSON object
    start = body.find("{")
    end = body.rfind("}")
    if start == -1 or end == -1 or end <= start:
        raise json.JSONDecodeError("No JSON object found", body, 0)
    return json.loads(body[start : end + 1])


def llm_review(
    original: str, 
    sanitized: str, 
    sim_score: float,
    model: str | None = None
) -> ReviewResult:
    """Phase 2: LLM reasoning for nuanced validation."""
    
    resolved_model = model or REVIEWER_MODEL
    
    messages = [
        {
            "role": "system", 
            "content": "You are a precise validator. Output ONLY valid JSON. No markdown, no explanation outside JSON."
        },
        {
            "role": "user",
            "content": _build_review_prompt(original, sanitized, sim_score)
        }
    ]
    
    try:
        response = chat(messages, model=resolved_model)
        result = _extract_json_from_response(response)
        
        # Validate structure
        checks = result.get("checks", {})
        approved = result.get("approved", False)
        
        return ReviewResult(
            approved=approved and all(checks.values()),
            confidence=result.get("confidence", 0.5),
            similarity_score=sim_score,
            checks=checks,
            missing_items=result.get("missing_items", []),
            reason=result.get("reason", ""),
            suggestions=[result.get("retry_suggestion", "")]
        )
        
    except (json.JSONDecodeError, ValueError) as e:
        return ReviewResult(
            approved=False,
            confidence=0.0,
            similarity_score=sim_score,
            checks={"llm_parse": False},
            missing_items=[],
            reason=f"LLM output not valid JSON: {str(e)[:100]}",
            suggestions=["Review prompt formatting"]
        )


def review_agent3(
    original: str,
    agent2_output: Any,  # Agent2Result or dict
    use_llm: bool = True,
    min_similarity: float = 0.5
) -> ReviewResult:
    """
    Main entry for Agent 3 review.
    
    Flow: Quick checks → (optional) LLM review
    """
    sanitized = (
        agent2_output.sanitized_prompt 
        if hasattr(agent2_output, 'sanitized_prompt')
        else agent2_output.get("sanitized_prompt", "")
    )
    
    # Phase 1: Quick deterministic checks
    quick_pass, quick_fail = _quick_checks(original, sanitized)
    if not quick_pass:
        return quick_fail
    
    sim = compute_similarity(original, sanitized)
    
    # Phase 2: LLM for nuanced checks
    if use_llm:
        return llm_review(original, sanitized, sim)
    
    # No LLM mode: rule-based final decision
    return ReviewResult(
        approved=sim >= min_similarity,
        confidence=0.7 if sim >= min_similarity else 0.3,
        similarity_score=sim,
        checks={"similarity_threshold": sim >= min_similarity},
        missing_items=[] if sim >= min_similarity else ["elements potentially lost"],
        reason=f"Similarity {sim:.3f} vs threshold {min_similarity}" 
               + ("" if sim >= min_similarity else " - possible over-compression"),
        suggestions=["Enable LLM review for detailed analysis"] if sim < 0.7 else []
    )

File: backend/src/agents/lib/reviewer.ts
/**
 * Agent 3: review / validation (ported from Hackathon/agents/reviewer.py).
 */

import { chat } from "./llm-chat";
import type { Agent2Result } from "./context-simplifier";
import {
  AGE_RE,
  EMAIL_RE,
  FILENAME_PATTERN,
  LOCATION_RE,
  NAME_RE,
  PHONE_RE,
  reTest,
} from "./patterns";
import { computeSimilarity } from "./similarity";

export interface ReviewResult {
  approved: boolean;
  confidence: number;
  similarityScore: number;
  checks: Record<string, boolean>;
  missingItems: string[];
  reason: string;
  suggestions: string[];
}

function asText(value: unknown): string {
  if (value == null) return "";
  if (typeof value === "string") {
    const text = value.trim();
    if (text.startsWith("{") && text.endsWith("}")) {
      try {
        const parsed = JSON.parse(text) as Record<string, unknown>;
        return asText(
          parsed.reason ?? parsed.explanation ?? parsed.message ?? parsed.summary ?? parsed,
        );
      } catch {
        return text;
      }
    }
    return text;
  }
  if (typeof value === "object" && !Array.isArray(value)) {
    const obj = value as Record<string, unknown>;
    for (const key of ["reason", "explanation", "message", "summary", "detail"]) {
      if (key in obj) return asText(obj[key]);
    }
    return Object.entries(obj)
      .map(([k, v]) => `${k}: ${asText(v)}`)
      .filter((s) => s.length > 0)
      .join("; ");
  }
  if (Array.isArray(value)) {
    return value.map(asText).filter(Boolean).join("; ");
  }
  return String(value).trim();
}

function asList(value: unknown): string[] {
  if (value == null) return [];
  if (Array.isArray(value)) return value.map(asText).filter(Boolean);
  const text = asText(value);
  return text ? [text] : [];
}

function normalizeChecks(value: unknown): Record<string, boolean> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return { llm_checks_valid: false };
  }
  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>).map(([k, v]) => [String(k), Boolean(v)]),
  );
}

function agent2Payload(agent2Output: Agent2Result | Record<string, unknown>): Record<string, unknown> {
  if ("sanitizedPrompt" in agent2Output && typeof agent2Output.sanitizedPrompt === "string") {
    const result = agent2Output as Agent2Result;
    const compressed = (result.compressed ?? {}) as Record<string, unknown>;
    return {
      sanitized_prompt: result.sanitizedPrompt,
      algorithm: result.algorithm || compressed.algorithm || "",
      compression_level: compressed.compression_level || "",
      active_files: compressed.active_files || [],
      errors: compressed.errors || [],
      constraints: compressed.constraints || [],
      open_questions: compressed.open_questions || [],
      important_symbols: compressed.important_symbols || [],
    };
  }

  const dict = agent2Output as Record<string, unknown>;
  const compressed = (dict.compressed ?? {}) as Record<string, unknown>;
  const source = typeof compressed === "object" && compressed !== null ? compressed : {};
  return {
    sanitized_prompt: dict.sanitized_prompt || source.sanitized_prompt || "",
    algorithm: dict.algorithm || source.algorithm || "",
    compression_level: dict.compression_level || source.compression_level || "",
    active_files: dict.active_files || source.active_files || [],
    errors: dict.errors || source.errors || [],
    constraints: dict.constraints || source.constraints || [],
    open_questions: dict.open_questions || source.open_questions || [],
    important_symbols: dict.important_symbols || source.important_symbols || [],
  };
}

function quickChecks(
  original: string,
  sanitized: string,
): { pass: boolean; fail?: ReviewResult } {
  if (!sanitized || sanitized.trim().length < 5) {
    return {
      pass: false,
      fail: {
        approved: false,
        confidence: 1,
        similarityScore: 0,
        checks: { empty: false },
        missingItems: ["sanitized_prompt"],
        reason: "Output is empty or too short",
        suggestions: ["Check Agent 2 pipeline"],
      },
    };
  }

  const piiPatterns = [EMAIL_RE, PHONE_RE, AGE_RE, NAME_RE, LOCATION_RE];
  const foundPii = piiPatterns.filter((p) => reTest(p, sanitized)).map((p) => p.source.slice(0, 20) + "...");

  if (foundPii.length > 0) {
    return {
      pass: false,
      fail: {
        approved: false,
        confidence: 1,
        similarityScore: 0,
        checks: { pii_clean: false },
        missingItems: foundPii,
        reason: "PII detected in output",
        suggestions: ["Increase PII masking in Agent 2"],
      },
    };
  }

  // Lexical cosine is low when rephrasing long→short; use lenient threshold
  const sim = computeSimilarity(original, sanitized);
  const lengthRatio = sanitized.length / Math.max(original.length, 1);
  // When heavily compressed (short output vs long input), lower the bar
  const threshold = lengthRatio < 0.3 ? 0.08 : 0.15;

  if (sim < threshold) {
    return {
      pass: false,
      fail: {
        approved: false,
        confidence: 0.9,
        similarityScore: sim,
        checks: { semantic_similarity: false },
        missingItems: ["core_intent"],
        reason: `Semantic drift too severe (${sim.toFixed(2)} < ${threshold})`,
        suggestions: ["Agent 2 over-compressed; retry with lower threshold"],
      },
    };
  }

  return { pass: true };
}

function extractKeyElements(text: string): Record<string, string[]> {
  return {
    filenames: [...text.matchAll(/\b\w+\.(py|js|ts|json|yaml|yml|toml|md|txt|csv|sql|html|css|scss|jsx|tsx)\b/gi)].map(
      (m) => m[0],
    ),
    errors: [...text.matchAll(/\b\w*Error:?\s*[^.\n]*/gi)].map((m) => m[0]),
    questions: [...text.matchAll(/[^.?!]*\?[^.?!]*/g)].map((m) => m[0]),
    constraints: [
      ...text.matchAll(
        /\b(do not|don't|must|should|never|always|keep|preserve|maintain)\b[^.?!]*[.?!]/gi,
      ),
    ].map((m) => m[0]),
    code_symbols: [...text.matchAll(/\b\w+\(\s*\)/g)].map((m) => m[0]),
    urls_apis: [...text.matchAll(/\b(api|endpoint|url|http)[^\s]*/gi)].map((m) => m[0]),
  };
}

function missingCriticalItems(original: string, sanitized: string): string[] {
  const originalElements = extractKeyElements(original);
  const sanitizedLower = sanitized.toLowerCase();
  const missing: string[] = [];

  for (const category of ["filenames", "errors", "code_symbols"] as const) {
    for (const item of originalElements[category] ?? []) {
      if (item && !sanitizedLower.includes(item.toLowerCase())) {
        missing.push(`${category}:${item}`);
      }
    }
  }

  for (const item of originalElements.constraints ?? []) {
    if (/\bdon't\s+(even\s+)?know\b/i.test(item)) continue;
    if (item && !sanitizedLower.includes(item.toLowerCase())) {
      missing.push(`constraints:${item}`);
    }
  }

  return missing;
}

function buildReviewPrompt(
  original: string,
  sanitized: string,
  simScore: number,
  agent2Meta: Record<string, unknown>,
): string {
  const origElems = extractKeyElements(original);
  const slice = (arr: string[], n: number) => arr.slice(0, n);

  return `Compare ORIGINAL vs SANITIZED and return ONE valid JSON object only.
Do not use markdown. Do not wrap the JSON in code fences. Do not add headings.

ORIGINAL (user's verbose prompt):
"""${original.slice(0, 2000)}"""

SANITIZED (simplified question by Agent 2):
"""${sanitized.slice(0, 1500)}"""

Pre-computed lexical similarity: ${simScore.toFixed(3)} (NOTE: low similarity is EXPECTED when rephrasing a long text into a short question)

Agent 2 metadata:
- Algorithm: ${agent2Meta.algorithm ?? ""}
- Compression level: ${agent2Meta.compression_level ?? ""}
- Important symbols: ${JSON.stringify(agent2Meta.important_symbols ?? [])}
- Constraints: ${JSON.stringify(agent2Meta.constraints ?? [])}

TASK: Agent 2's job is to SIMPLIFY the user's prompt into a concise question. It is NOT required to keep every detail.

Rules for approval:
1. CORE INTENT: Does the simplified question capture what the user actually wants to know/do?
2. KEY DECISION: If the user is asking "should I do X or Y", is that choice preserved?
3. CONSTRAINTS: Are hard requirements ("must", "don't", specific goals like "for internship demo") kept?
4. NOT REQUIRED to keep: background stories, names, locations, all technical details that are just context

Decision:
- APPROVE if the simplified question correctly identifies what the user needs answered
- APPROVE even if minor supporting details are dropped (that's the point of simplification)
- FAIL only if the core question/intent is WRONG or MISSING entirely

Return exactly this JSON shape:
{
  "approved": true,
  "confidence": 0.8,
  "checks": {
    "intent_preserved": true,
    "key_decision_preserved": true,
    "constraints_preserved": true
  },
  "missing_items": ["only list truly critical things that change the meaning"],
  "reason": "one plain-text sentence",
  "retry_suggestion": "one plain-text hint for Agent 2"
}`;
}

function extractJsonFromResponse(text: string): Record<string, unknown> {
  let body = text.trim();
  body = body.replace(/<think>[\s\S]*?<\/redacted_thinking>/gi, "").trim();

  const fenced = body.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
  if (fenced?.[1]) {
    body = fenced[1].trim();
  }

  const start = body.indexOf("{");
  const end = body.lastIndexOf("}");
  if (start === -1 || end === -1 || end <= start) {
    throw new SyntaxError("No JSON object found");
  }
  return JSON.parse(body.slice(start, end + 1)) as Record<string, unknown>;
}

function fallbackReview(
  original: string,
  sanitized: string,
  simScore: number,
  reason: string,
): ReviewResult {
  const missingItems = missingCriticalItems(original, sanitized);
  const approved = simScore >= 0.72 && missingItems.length === 0;
  return {
    approved,
    confidence: approved ? 0.65 : 0.35,
    similarityScore: simScore,
    checks: {
      llm_parse: false,
      semantic_similarity: simScore >= 0.72,
      critical_items_preserved: missingItems.length === 0,
    },
    missingItems,
    reason: `LLM reviewer returned malformed JSON; used deterministic fallback. ${reason}`,
    suggestions: approved ? [] : ["Retry LLM review or use a stronger reviewer model"],
  };
}

async function llmReview(
  original: string,
  sanitized: string,
  simScore: number,
  agent2Meta: Record<string, unknown>,
): Promise<ReviewResult> {
  const messages = [
    {
      role: "system" as const,
      content:
        "You are a precise validator. Output ONLY valid JSON. No markdown, no explanation outside JSON.",
    },
    {
      role: "user" as const,
      content: buildReviewPrompt(original, sanitized, simScore, agent2Meta),
    },
  ];

  try {
    const response = await chat(messages);
    const result = extractJsonFromResponse(response);
    const checks = normalizeChecks(result.checks);
    const approved = Boolean(result.approved);
    const reason = asText(result.reason);
    const suggestions = asList(result.retry_suggestion);

    return {
      approved: approved && Object.values(checks).every(Boolean),
      confidence: Number(result.confidence ?? 0.5),
      similarityScore: simScore,
      checks,
      missingItems: asList(result.missing_items),
      reason,
      suggestions,
    };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return fallbackReview(original, sanitized, simScore, `Parse error: ${msg.slice(0, 100)}`);
  }
}

export async function reviewAgent3(
  original: string,
  agent2Output: Agent2Result | Record<string, unknown>,
  options: { useLlm?: boolean; minSimilarity?: number } = {},
): Promise<ReviewResult> {
  const { useLlm = true, minSimilarity = 0.5 } = options;
  const agent2Meta = agent2Payload(agent2Output);
  const sanitized = asText(agent2Meta.sanitized_prompt);

  const quick = quickChecks(original, sanitized);
  if (!quick.pass && quick.fail) {
    return quick.fail;
  }

  const sim = computeSimilarity(original, sanitized);

  if (useLlm) {
    return llmReview(original, sanitized, sim, agent2Meta);
  }

  const approved = sim >= minSimilarity;
  return {
    approved,
    confidence: approved ? 0.7 : 0.3,
    similarityScore: sim,
    checks: { similarity_threshold: approved },
    missingItems: approved ? [] : ["elements potentially lost"],
    reason:
      `Similarity ${sim.toFixed(3)} vs threshold ${minSimilarity}` +
      (approved ? "" : " - possible over-compression"),
    suggestions: sim < 0.7 ? ["Enable LLM review for detailed analysis"] : [],
  };
}
